src/main/java/entity_layer/RandomPasswordGenerator.java

```diff
@@ -0,0 +1,160 @@
+     * @return a random number of special characters in a char array
+     */
+    private char[] getSpecialCharacters() {
+        int numSpecial = this.getRandomInt(1, this.getUpperBound());
+        this.numCharsToAdd -= numSpecial;
+
+        return this.getRandomizedChars(numSpecial, 33, 48);
+    }
+
+    /**
+     * Gets the number characters to put in the new random password
+     * @return a random number of number characters in a char array
+     */
+    private char[] getNumberCharacters() {
+        int numberChars = this.getRandomInt(1, this.getUpperBound());
+        this.numCharsToAdd -= numberChars;
+
+        return this.getRandomizedChars(numberChars, 48, 58);
+    }
+
+    /**
+     * Gets the maximum number of the random int to select + 1, since the upperBound of the getRandomInt function is
+     * non-inclusive
+     * @return an integer representing the maximum number of characters that can be selected + 1
+     */
+    private int getUpperBound() {
+        int IRREGULAR_CHAR_PROPORTION = 3;
+        return Math.max(2, this.numCharsToAdd / IRREGULAR_CHAR_PROPORTION);
+    }
+
+    /**
+     * Gets the number of regular letter characters to add to this password, using the remaining number of characters
+     * to obtain the requested length
+     * @return lower case letters in a char array
+     */
+    private char[] getNormalCharacters() {
+        return this.getRandomizedChars(this.numCharsToAdd, 97, 123);
+    }
+
+    /**
+     * Gets a random int using an IntStream that is greater than or equal to lowerBound and less than upperBound
+     * If the bounds are invalid, the lowerBound is returned
+     * @param lowerBound the lowest possible int returned
+     * @param upperBound the highest possible int returned + 1
+     * @return a random int between lower and upper bound values
+     */
+    private int getRandomInt(int lowerBound, int upperBound) {
+        if (lowerBound < upperBound) {
+            IntStream options = randomizer.ints(lowerBound, upperBound);
+            return options.limit(1).toArray()[0];
+        } else {
+            return lowerBound;
+        }
+    }
+
+    /**
+     * Gets numRequested number of random characters with ASCII values between the upper and lower bounds
+     * @param numRequested number of characters to return
+     * @param lowerBound the lowest ASCII value of the characters generated
+     * @param upperBound the highest possible ASCII value of the characters generated + 1
+     * @return a char array of length numRequested, where all ASCII values of chars are between lower and upper bound
+     */
+    private char[] getRandomizedChars(int numRequested, int lowerBound, int upperBound) {
+        IntStream lengthOptions = randomizer.ints(lowerBound, upperBound);
+        int [] chosenInts = lengthOptions.limit(numRequested).toArray();
+        char[] chosenChars = new char[numRequested];
+
+        for (int i = 0; i < numRequested; i++) {
```

Comment:
Really creative use of for loop to get the values of the characters between the upper ad lower bounds, this will be really helpful with the recursion of my tasktree UI.